### ISSUE TITLE
brace-style "allowSingleLine" should differentiate between if/else and function statements.

### ISSUE THREAD

**Issue by Droogans** [eslint/eslint#5592]
**What version of ESLint are you using?**

v2.2.0

**What parser (default, Babel-ESLint, etc.) are you using?**

Default.

**Please show your full configuration:**

```js
"brace-style": [2, "1tbs", { "allowSingleLine": true }]
```

**What did you do? Please include the actual source code causing the issue.**

I have two snippets that eslint is (correctly) reporting. However, I'd like to disable half of the violations.

> good, please continue flagging as an error
```js
if (foo) { doFoo(); } else { doOther(); }
```

> please allow me to disable errors for these statements
```js
var o = { action: function () { doSomething(); } };
```

**What did you expect to happen?**

I expected `"allowSingleLine"` to accept an object to expose more granularity in my config.

```js
"brace-style": [2, "1tbs", { "allowSingleLine": { "ifStatements": false, "functions": true } }]
```

**What actually happened? Please include the actual, raw output from ESLint.**

I'm seeing the usual errors being reported. Nothing out of the ordinary.

```
error  Statement inside of curly braces should be on next line   brace-style
```

**Comment by platinumazure** [comment]
Just to be clear, this is a request for new functionality on the `brace-style` rule, correct? The rule currently does not support what you are asking for.

Could you flesh out your proposal a little more with some other use cases? You mentioned if statements and functions-- would this apply to any statement type that has one or more BlockStatement nodes underneath it, where finer granularity would be allowed?

Does this proposal make sense on all brace-style options, or only in 1tbs?

**Comment by Droogans** [comment]
> this is a request for new functionality on the brace-style rule, correct?

This is most certainly a recommendation for enhancing the existing `"braceStyle"/"allowSingleLine"` rule.

> Does this proposal make sense on all brace-style options, or only in 1tbs?

I would think that if every brace-style pre-set ("allman", "stroustrup", etc.) accepts a "singleLine" parameter, then it makes sense to allow for this proposed extension to be used in all of them.

> Could you flesh out your proposal a little more with some other use cases?

Sure. This is what I want to write.

```js
"brace-style": [2, "1tbs", { "allowSingleLine": { "ifStatements": false, "functions": true } }]
```

And some javascript to show what this might look like.

```js
if (!_.isPlainObject(someThing)) {
    someThing = { updateSomething: function () { return somethingElse.update(); } }; // this is not lint
} else {
    someThing.updateSomething = function () {
        return somethingOtherThanElse.update();
    };
}
```

If I were to write a one-liner `if`/`else` statement, it would throw an error.

I should probably link to another issue I've raised here, #5520, and add that there's already one request by me to differentiate between functions and getters/setters, so that may make sense here too.

```js
"brace-style": [2, "1tbs", { "allowSingleLine": {
    "ifStatements": false,
    "functions": false,
    "gettersSetters": true
}}]
```

That would have this kind of code.

```js
var o = {
    get prop() { return 'a'; }, // OK
    shortFn: function () { return 'b'; }, // not OK
    longerFn: function () {
        return 'c'; // OK
    }
};
```

**Comment by BYK** [comment]
Not really a fan but I can definitely understand your point of view. Guess `// eslint-disable-line brace-style` doesn't work for you since it is too common in the code?

**Comment by Droogans** [comment]
If this issue is accepted, fixed, and released, then I'd use it.

If you don't? I'd spend an hour or so fixing a few dozen instances of the "no single line functions" offenses. Not a huge deal.

Just something to think about. I don't have strong feeling one way or the other.

**Comment by platinumazure** [comment]
Thanks for fleshing this out @Droogans. On the one hand, I can see why this would be useful; on the other hand, I'm worried about the maintainability of the rule-- it would get a lot more complex.

Can you answer @BYK's question about how often it appears in the code? Would `// eslint-disable-line brace-style` work on a case-by-case basis?

**Comment by Droogans** [comment]
> Would `// eslint-disable-line` brace-style work on a case-by-case basis?

I'd much rather revert all one-liner functions to multi-line to avoid that issue entirely. I would go mad trying to keep an entire team's worth of contributions using both single line functions and this comment above them all!

**Comment by alberto** [comment]
Thanks for your interest in improving eslint. Unfortunately, it looks like consensus couldn't be reached on this issue and so I'm closing it. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to reach consensus after 21 days tend never to reach consensus, and as such, we close those issues. This doesn't mean the idea isn't interesting, just that it's not something the team can commit to.

As a reminder, you can still create a custom rule for your use use.
